### ISSUE TITLE
absolute vs relative mount targets

### ISSUE THREAD

**Issue by flx42** [lxc/lxc#2276]
When using `lxc.mount.entry`, two very different code paths will be triggered depending on whether the target is [absolute](https://github.com/lxc/lxc/blob/master/src/lxc/conf.c#L2182-L2224) or [relative](https://github.com/lxc/lxc/blob/master/src/lxc/conf.c#L2226-L2240).

Unless I've missed it, this distinction doesn't seem to be clearly documented. And I must admit I don't fully understand what the `absolute` code path is trying to do. Especially, why does it append lxcpath and the container name? That's actually not where my rootfs is.

**Comment by flx42** [comment]
My understanding today is that an absolute path means absolute on the host filesystem. Looks like the target **needs** to include either `<lxcpath>/<name>/rootfs/` or `<lxc.rootfs.path>/`.
Since mounts outside of these paths are rejected, shouldn't we treat absolute and relative targets the same way? e.g., relative to the container root mount point.

By the way, the logic is broken today, `strstr` is not what we want. This will work fine:
```
lxc.rootfs.path = rootfs
lxc.mount.entry = /home/fabecassis /a/b/c/d/myrootfsfoo/myhome bind rbind,rprivate,create=dir 0 0
```
The target is totally bogus, but happens to have `rootfs` as a substring, so LXC will happily accept it and mount the folder at `<mountpoint>/foo/myhome`

**Comment by brauner** [comment]
@flx42, I think the logic that we want is that absolute and relative targets are identical.

**Comment by hallyn** [comment]
@brauner no,  long long ago the behavior was indeed meant to be "absolute pathname is absolute relative to the host."  So I think we need to keep supporting that, whacky though it may be.  I think it made more sense in the days when an lxc application container could not have a rootfs of its own, i.e. share /.
So I think fixing the strstr is the only action here.

**Comment by flx42** [comment]
And I guess adding a new configuration item with an unified behavior is not an option?

**Comment by brauner** [comment]
@flx42, @hallyn, in that case we should:

- interpret absolute paths the same as relative paths if the container has a separate rootfs
- interpret absolute paths as absolute when the container has no rootfs path

opinions?

**Comment by hallyn** [comment]
Quoting Christian Brauner (notifications@github.com):
> @flx42, @hallyn, in that case we should:
> 
> - interpret absolute paths the same as relative paths if the container has a separate rootfs

except when the absolute path is the rootfs path, as I'm sure
there are a lot of places still using that format, i.e.

lxc.mount.entry = /path /var/lib/lxc/container/rootfs/path ...

**Comment by flx42** [comment]
> interpret absolute paths the same as relative paths if the container has a separate rootfs

It would make sense, I think. But it would indeed be a breaking change as @hallyn pointed out.
Not sure what is your policy regarding deprecating/breaking the syntax of the configuration file.

**Comment by brauner** [comment]
> except when the absolute path is the rootfs path, as I'm sure
there are a lot of places still using that format, i.e.

> lxc.mount.entry = /path /var/lib/lxc/container/rootfs/path ...

Then you need something like:

```
target = "/path /var/lib/lxc/container/rootfs/path";
if (has_rootfs)
        /* treat as absolute */
else if (target[0] = '/') {
        cleanpath = lxc-deslashify("/path /var/lib/lxc/container/rootfs/path");
        if (strncmp(cleanpath, rootfs_path, strlen(rootfs_path))
                /* handle as absolute path */
        else
                /* handle as relative path */
```

@hallyn, @flx42, does that make sense?

**Comment by flx42** [comment]
This could make sense in the path for deprecating the previous behaviour, towards what you suggested in https://github.com/lxc/lxc/issues/2276#issuecomment-381427396. But it seems too confusing to keep this special case forever.

**Comment by brauner** [comment]
Actually, I think that the current behavior is buggy and does not really promise what it is doing. Let me test.

**Comment by brauner** [comment]
1. `lxc.mount.entry = /home/chb/files/Cground opt none bind,create=dir 0 0`
```
lxc-execute xenial 20180415210659.121 DEBUG    lxc_conf - conf.c:mount_entry:1955 - Remounting "/home/chb/files/Cground" on "/usr/lib/x86_64-linux-gnu/lxc/opt" to respect bind or remount options
lxc-execute xenial 20180415210659.121 DEBUG    lxc_conf - conf.c:mount_entry:1976 - Flags for "/home/chb/files/Cground" were 4096, required extra flags are 0
lxc-execute xenial 20180415210659.121 DEBUG    lxc_conf - conf.c:mount_entry:1986 - Mountflags already were 4096, skipping remount
lxc-execute xenial 20180415210659.121 DEBUG    lxc_conf - conf.c:mount_entry:2032 - Mounted "/home/chb/files/Cground" on "/usr/lib/x86_64-linux-gnu/lxc/opt" with filesystem type "none"
```
2. `lxc.mount.entry = /home/chb/files/Cground /opt none bind,create=dir 0 0`
```
lxc-execute xenial 20180415210750.472 WARN     lxc_conf - conf.c:mount_entry_on_absolute_rootfs:2213 - Ignoring mount point "/opt"
lxc-execute xenial 20180415210750.472 INFO     lxc_conf - conf.c:mount_file_entries:2264 - Finished setting up mounts
```

3. `lxc.mount.entry = /home/chb/files/Cground /home/chb/.local/share/lxc/xenial/rootfs/opt none bind,create=dir 0 0`
```
lxc-execute xenial 20180415210902.482 DEBUG    lxc_conf - conf.c:mount_entry:1955 - Remounting "/home/chb/files/Cground" on "/usr/lib/x86_64-linux-gnu/lxc//opt" to respect bind or remount options
lxc-execute xenial 20180415210902.482 DEBUG    lxc_conf - conf.c:mount_entry:1976 - Flags for "/home/chb/files/Cground" were 4096, required extra flags are 0
lxc-execute xenial 20180415210902.482 DEBUG    lxc_conf - conf.c:mount_entry:1986 - Mountflags already were 4096, skipping remount
lxc-execute xenial 20180415210902.482 DEBUG    lxc_conf - conf.c:mount_entry:2032 - Mounted "/home/chb/files/Cground" on "/usr/lib/x86_64-linux-gnu/lxc//opt" with filesystem type "none"
```
I don't terribly mind special-casing case 3 but we should change the behavior for the other case where it is not prefixed by the path to the rootfs.

**Comment by stgraber** [comment]
@brauner is that something we should sort out for 4.0?
